streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala

```diff
@@ -44,6 +49,14 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
   private val conf = ssc.conf
   private val graph = ssc.graph
 
+  // Scala treeset makes it difficult to drop items while iterating over the set. dropWhile
+  // creates a new set, which does not seem very efficient. So use Java treeset instead.
+  private val activeBatches =
+    Collections.synchronizedSet[Time](new TreeSet[Time](new TimeComparator()))
```

Comment:
Scala treeset does not have a way of dropping data based on some predicate. `dropWhile` seems inefficient since it creates a new set - which seemed unnecessary. An alternative would be using the iterator to drop, but the scala iterator does not allow removing data.